javalib/src/main/scala/java/lang/Throwables.scala

```diff
@@ -49,49 +49,126 @@ private[lang] object StackTrace {
+
   def this(s: String) = this(s, null)
+
   def this(e: Throwable) = this(if (e == null) null else e.toString, e)
 
-  private var stackTrace: Array[StackTraceElement] = _
+  private[this] var stackTrace: Array[StackTraceElement] = _
 
-  fillInStackTrace()
+  if (writableStackTrace)
+    fillInStackTrace()
 
-  def initCause(cause: Throwable): Throwable = {
-    e = cause
-    this
-  }
+  // We use an Array rather than, say, a List, so that Throwable does not
+  // depend on the Scala collections.
+  private[this] var suppressed: Array[Throwable] = _
 
-  def getMessage(): String = s
+  final def addSuppressed(exception: Throwable): Unit = {
 
-  def getCause(): Throwable = e
+    if (exception == null) {
+      // Yes, JVM message uses 'cannot' (sic) & has terminal period/full_stop!
+      throw new java.lang.NullPointerException(
+        "Cannot suppress a null exception.")
+    }
 
-  def getLocalizedMessage(): String = getMessage()
+    if (exception == this) {
+      throw new java.lang.IllegalArgumentException(
+        "Self-suppression not permitted")
+    }
+
+    if (enableSuppression) this.synchronized {
+      if (suppressed eq null) {
+        suppressed = Array(exception)
+      } else {
+        val length        = suppressed.length
+        val newSuppressed = new Array[Throwable](length + 1)
+        System.arraycopy(suppressed, 0, newSuppressed, 0, length)
+        newSuppressed(length) = exception
+        suppressed = newSuppressed
+      }
+    }
+  }
 
   def fillInStackTrace(): Throwable = {
-    this.stackTrace = StackTrace.currentStackTrace()
+    // currentStackTrace should be handling exclusion in its own
+    // critical section, but does not. So do
+    if (writableStackTrace) this.synchronized {
+      this.stackTrace = StackTrace.currentStackTrace()
+    }
     this
   }
 
+  def getCause(): Throwable = e
+
+  def getLocalizedMessage(): String = getMessage()
+
+  def getMessage(): String = s
+
   def getStackTrace(): Array[StackTraceElement] = {
-    if (stackTrace eq null) {
-      stackTrace = Array.empty
-    }
-    stackTrace
+    // Be robust! Test this.stackTrace against null rather than relying upon
+    // the value of writableStackTrace.
+    //
+    // subclass scala.util.control.NoStackTrace overrides fillInStackTrace()
+    // so that it never touches this.stackTrace. This creates the situation
+    // where writeableStackTrace is true and this.stackTrace is null.
+    //
+    // If scala code creates this situation, then by the Gell-Mann principle
+    // user code in the wild is bound to do so.
+    //
+    // If stackTrace is null, no profit to calling fillInStackTrace().
+    // If writableStackTrace is true, then fillInStackTrace has already
+    // been called in the constructor. If the stack is not writable, then
+    // it can not be filled in.
+
+    if (stackTrace == null) {
+      Array.empty[StackTraceElement] // as specified by Java 8.
+    } else
+      this.synchronized {
```

Comment:
AFAICT, this `synchronized` is unnecessary.

Reply:
I think an umpire's call is needed here.  I will follow your recommendation/requirement.

I believe that Array clone is not-thread-safe.  At the least, I can add a comment 
to the `suppressed` variable that it is not thread-safe and any given Throwable should
only be used by in a single thread.  If I were taking a Concurrency exam, that would
at least show that I had recognized the issue (and then punted it to future devos).

Scala Native currently follows the Scala.js  single threading model. 
I believe the synchronized is currently a call into a monitor method implemented as
a noop.  So, there is some execution cost and some review cognitive cost.

From my Operating System network kernel days, I tend to write thread-safe code
from the beginning because it is almost impossible to go back later and find the
non-thread-safe parts.   

It may be years in the future, but someday there will be customer pull for SN to
be multi-threaded.  I concur that currently SN has higher priority issues than
investing huge amount of time in thread safety.

I also believe that it would be a perverse & probably stupid multi-threaded
program which tried to write the same (eq) Throwable from two or more threads.


Large parts of the java.concurrent package are only EventLoop safe, and are 
documented in the code as such.  I do not remember if the ReadTheDocs documents
describe them as such.  I could check.

Reply:
I don't mean to say it's not necessary because synchronization is not necessary in general (otherwise I'd have recommended to remove *all* `synchronized`). But in this case I'm pretty sure it's not necessary.

`clone()` is not thread-safe while mutating the original array. However, any array that is stored in the field `stackTrace` is never modified after it is stored there. Therefore, when you read an array from `stackTrace`, it is effectively immutable, and you can `clone()` it without synchronization.

Reply:
Thank you for the patient explanation.  Changed as recommended.